lib/puppet/pops/patterns.rb

```diff
@@ -48,7 +48,8 @@ module Puppet::Pops::Patterns
 
   # VAR_NAME matches the name part of a variable (The $ character is not included)
   # Note, that only the final segment may start with an underscore.
-  VAR_NAME = %r{\A(?:((::)?[a-z]\w*)*(?:(::)?_\w*)?)\z}
+  # Note, regexp sensitive to backtracking - hence its complexity
+  VAR_NAME = %r{\A(?:::){0,1}(?:(?:[a-z_]\w*)|(?:(?:(?:(?!_)[a-z]\w*))::)+(?:[a-z_]\w*))\z}
```

Comment:
I tested a much simpler variant of this regexp which seems to be just as performant:
```
VAR_NAME = %r{\A(?:::)?[a-z]\w*(?:::[a-z]\w*)*(?:::[a-z_]\w*)?\z}
```